File: TP3/JAVA/src/test/resources/TestObserver.java
package test.resources;

import org.junit.jupiter.api.Test;

import main.resources.ScolariteManager;
import src.Etudiant;

import static org.junit.jupiter.api.Assertions.*;

public class TestObserver {
    
    @Test
    def testObserverNotification() {
        ScolariteManager manager = ScolariteManager();
        
        Etudiant alice = Etudiant("Alice", 20, "E001", "Python");
        manager.ajouterEtudiant(alice);
        alice.ajouterNote(15);
        
        assertEquals(alice.getMoyenne(), 15.0);
        assertEquals(alice.observateur, manager);
    }
}

File: TP4/PYTHON/test/test_observer.py
from src.codes_initiaux.etudiant import Etudiant
from src.pattern.Singleton.ScolariteManager import ScolariteManager

def test_observer_notification():
    manager = ScolariteManager()
    
    alice = Etudiant("Alice", 20, "E001", "Python")
    manager.ajouter_etudiant(alice)
    
    alice.ajouter_note(15)
    
    assert len(alice.notes) == 1
    assert alice.get_moyenne() == 15.0
    assert manager in alice._observers